### ISSUE TITLE
need help with config and training algorithm

### ISSUE THREAD

**Issue by Kugelkopf** [Unity-Technologies/ml-agents#1791]
Hi everyone!
I would need help in 2 issues, but hopefully the solution to the 1st one could get me closer to the 2nd one.
1.)I have re-created the 3D Ball environment but only with 1 platform. I trained it through 500k iterations , but it learned slowly and it could never get pass the mean reward 100. In  the end it almost always got 100 as mean reward but never could get passed by it. This was my training configuration for that:
INFO:mlagents.envs:Hyperparameters for the PPO Trainer of brain Brain:
        batch_size:     1024
        beta:   0.005
        buffer_size:    10240
        epsilon:        0.2
        gamma:  0.99
        hidden_units:   128
        lambd:  0.95
        learning_rate:  0.0003
        max_steps:      5.0e5
        normalize:      False
        num_epoch:      3
        num_layers:     2
        time_horizon:   64
        sequence_length:        64
        summary_freq:   1000
        use_recurrent:  False
        graph_scope:
        summary_path:   ./summaries/ppo-0
        memory_size:    256
        use_curiosity:  False
        curiosity_strength:     0.01
        curiosity_enc_size:     128
Which parameters should be changed in order to achieve better performance?
2.) I have a very similar problem like in this document (Learning-Environment-Create-New.md),
though my issue is a little bit different. 
In my environment the agent is also a box-like object and it is trying to find another box-like object.
- I am using raycasting from 6 positions for my agent and the environment has 2 tags : "obstacle" and "goal"
- There are many obstacles and one goal, and the agent need to find the goal by going around in the map trying to find the goal, while avoiding the obstacles.
- The goal changes places randomly for 3 different positions. (Trying to avoid overfitting)
-Im only feeding my agent these informations:

    public override void CollectObservations()
    {     
            AddVectorObs(GetStepCount() / (float)agentParameters.maxStep);
          //observing distances with the objects
            AddVectorObs(rayPer.Perceive(rayDistance, rayAngles, detectableObjects, 0f, 0f));       
            AddVectorObs(gameObject.transform.rotation.z);
            AddVectorObs(gameObject.transform.rotation.x);
            AddVectorObs(agentRb.transform.position.x - gameObject.transform.position.x);
            AddVectorObs(agentRb.transform.position.z - gameObject.transform.position.z);
            AddVectorObs(agentRb.transform.position.y - gameObject.transform.position.y);
            AddVectorObs(agentRb.velocity);
    }
- The agents gets rewarded (1.0f) when colliding with the "goal"
- The agent gets a negative reward (-1.0f) when the number of max steps are done which is 10k now
- The agent gets a negative reward (-1.0f) when colliding with an "obstacle"

I would like my agent to learn to avoid the obstacles and move efficiently to find the "goal" which could be at any place of the 3 random positions. I would be glad to know which parameters should be changed in what manner in order to achive it. Also I would like to try it make it work with a customized DQN  algorithm if it's reasonable or maybe I should customize the given PPO.

Im happy for any kind of help. The 2nd issue would be my primer goal to achieve.

**Comment by TashaSkyUp** [comment]
Seems like it might be a sparse reward problem. What hyper parameters have
you tried? Personally I would try calculating a small reward per decision
step based on how close it is to reaching the goal.. Train it until it
understands how to reach the goal and can generalize then slowly remove the
artificial reward. Also it may be worth trying the other way around..
Slowly move the start location further away from the goal each time it
finds it.

**Comment by Kugelkopf** [comment]
Hi! 
Thank you for your reply!
I used the same configuration like in the 1st issue I posted. I was thinking I would not use any position-based rewarding or anything which could lead the agent to learn "making the right steps at the right time". I'm saying this because yesterday when I ran the PPO with the above mentioned configs, the agent actually seemed to be learning that turning to the left should reward it sometimes. ( As the "goal" object appears at one of the 3 random positions everytime the environment resets) , because the "goal" object appears 2 out of 3 times on the left, and those positions are not really covered by "obstacles".

What I would like my agent to actually learn is to identify the "obstacles" and avoid them, and to identify the "goal" object and search for it. I would like my agent to have an "inner" motive to search around the map while avoiding the obstacles which it learns by colliding with them, and then learning that colliding with the "goal" object should be its goal. 

I believe its a mixture of the Sample environments or just a bit more complex problem than them, but Im a bit stuck at this point.
